### PR TITLE
Decode multi edge

### LINKED ISSUE #825: Improve decode time
When loading large RDBs files which include multi-edges loading time increases dramatically.
This due to naive handling of multi-edge decoding / processing.

### PR DESCRIPTION
**Pull request by swilly22**
fixes #825

## Release Notes

- **New Features**
  - Introduced support for version 16 in serialization and deserialization processes, enhancing graph data handling.
  - Added functionality for batch processing of graph connections, optimizing performance.
  - Implemented new test cases for multiple decoder versions, improving test coverage.
  - Enhanced graph entity serialization with new functions for encoding nodes and edges.
  - Added a new function to return the number of nodes the graph can contain, improving graph capacity management.
  - Enhanced the testing framework for upgrade scenarios using Docker containers.
  - Improved handling of edge loading and graph schema serialization, ensuring compatibility with version 16.
  - Added a new configuration parameter, `DELAY_INDEXING`, allowing for more control over indexing behavior during graph decoding.

- **Bug Fixes**
  - Improved error handling and validation in various serialization functions.

- **Documentation**
  - Updated comments and formatting for clarity across multiple files.

- **Tests**
  - Streamlined and expanded test cases for graph copying and persistency, ensuring robustness against various scenarios.
  - Adjusted test logic for better focus and clarity in graph copying functionality.
  - Updated test cases to reflect changes in naming conventions for nodes and edges.
  - Enhanced output and structure in graph comparison tests for better debugging.
  - Integrated upgrade tests into the testing framework, allowing for comprehensive testing of upgrade processes.
  - Updated configuration tests to verify the new `DELAY_INDEXING` parameter and the total number of configurations.

### CODE REVIEW

File: src/graph/tensor/tensor.c [hunk 1]
```diff
@@ -96,8 +96,187 @@ enum SetMethod {
 // set multiple entries
 void Tensor_SetElements
```
Comment (AviAvni) [inline src/graph/tensor/tensor.c:97]: what is the diff between Tensor_SetEdges and this function?

File: src/serializers/decoders/current/v16/decode_graph_entities.c [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+	SIValue list = SI_Array(arrayLen);
+	for(uint i = 0; i < arrayLen; i++) {
+		SIValue elem = _RdbLoadSIValue(rdb);
+		SIArray_Append(&list, elem);
+		SIValue_Free(elem);
+	}
+	return list;
+}
+
+static SIValue _RdbLoadVector
```
Comment (AviAvni) [inline src/serializers/decoders/current/v16/decode_graph_entities.c:75]: it look like we don't test loading vector from rdb

File: src/serializers/decoders/current/v16/decode_graph_entities.c [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+		Serializer_Graph_AllocEdgeAttributes(gc->g, e.id, &e);
+		_RdbLoadEntity(rdb, gc, (GraphEntity *)&e);
+
+		// index edge
+		if(PENDING_IDX(s)) {
+			Index_IndexEdge(PENDING_IDX(s), &e);
+		}
+
+		// update max node id
+		max_id = MAX(max_id, MAX(e.src_id, e.dest_id));
```
Comment (AviAvni) [inline src/serializers/decoders/current/v16/decode_graph_entities.c:407]: what we do with this max_id?

File: src/serializers/decoders/prev/v10/decode_graph_entities.c [hunk 4]
```diff
@@ -125,32 +130,149 @@ void RdbLoadDeletedNodes_v10(RedisModuleIO *rdb, GraphContext *gc, uint64_t dele
...
+		// flush batches
+		//----------------------------------------------------------------------
+
+		// flush batch when:
+		// 1. batch is full
+		// 2. relation id changed
+		if(idx > 0 && (idx >= BATCH_SIZE || relation_changed)) {
+			// flush batch
+			Serializer_OptimizedFormConnections(gc->g, prev_relation, srcs,
+					dests, ids, idx, max_node_id, false);
```
Comment (AviAvni) [inline src/serializers/decoders/prev/v10/decode_graph_entities.c:213]: take a look at all the uncovered lines in all decoders version

**Review by AviAvni — CHANGES_REQUESTED** [review]

File: tests/flow/test_prev_rdb_decode.py [hunk 5]
```diff
@@ -1,12 +1,267 @@
...
+        self.env.assertEqual(results.result_set, [[node0, edge01, node1]])
+
+        plan = str(graph.explain("MATCH (n:L1 {val:1}) RETURN n"))
+        self.env.assertIn("Index Scan", plan)
+
+        results = graph.query("MATCH (n:L1 {val:1}) RETURN n")
+        self.env.assertEqual(results.result_set, [[node0]])
+
+    GRAPH_ID = "v9_rdb_restore"
+    def _test_v9_decode(self):
```
[the remark this answers is not in the document]
Comment (swilly22) [reply tests/flow/test_prev_rdb_decode.py:264]: 👍

File: src/serializers/decoders/current/v16/decode_graph_schema.c [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+static void _RdbDecodeIndexField
+(
+	SerializerIO rdb,
+	char **name,             // index field name
+	IndexFieldType *type,    // index field type
+	double *weight,          // index field option weight
+	bool *nostem,            // index field option nostem
+	char **phonetic,         // index field option phonetic
+	uint32_t *dimension,     // index field option dimension
+	size_t *M, 		         // index field option M
```
[the remark this answers is not in the document]
Comment (swilly22) [reply src/serializers/decoders/current/v16/decode_graph_schema.c:18]: 👍

File: src/serializers/graph_extensions.c [hunk 7]
```diff
@@ -129,84 +129,100 @@ void Serializer_Graph_SetNodeLabels
...
+		ASSERT(info == GrB_SUCCESS);
+
+		//----------------------------------------------------------------------
+		// update relationship matrix
+		//----------------------------------------------------------------------
+
+		if(!multi_edge) {
+			//uint64_t _x;
+			//info = GrB_Matrix_extractElement_UINT64(&_x, m, row, col);
+			//ASSERT(info == GrB_NO_VALUE);
```
[the remark this answers is not in the document]
Comment (swilly22) [reply src/serializers/graph_extensions.c:152]: 👍

File: src/serializers/decoders/current/v16/decode_graph_entities.c (hunk 3 above)
Comment (swilly22) [reply src/serializers/decoders/current/v16/decode_graph_entities.c]: Removed, I don't believe it is needed.

**Review by AviAvni — CHANGES_REQUESTED** [review]

File: src/serializers/decoders/current/v16/decode_graph_entities.c [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+	for(uint64_t i = 0; i < node_count; i++) {
+		Node n;
+		NodeID id = SerializerIO_ReadUnsigned(rdb);
+
+		// #labels M
+		uint64_t nodeLabelCount = SerializerIO_ReadUnsigned(rdb);
+
+		// (labels) x M
+		LabelID labels[nodeLabelCount];
+		for(uint64_t i = 0; i < nodeLabelCount; i ++){
```
[the remark this answers is not in the document]
Comment (swilly22) [reply src/serializers/decoders/current/v16/decode_graph_entities.c]: 👍

**Pull request merged**